### ISSUE TITLE
Possible bug in DecoderRNN class

### ISSUE THREAD

**Issue by sohaib023** [HHTseng/video-classification#42]
Hello @HHTseng,

In DecoderRNN both input and output of the self.LSTM layer has dim=0 as batch dimension and dim=1 as timestep dimension. However, the pytorch documentation (link provided below) has those two dimensions reversed both for the input and output. 
Is there any specific reason for this difference in implementation? If not, this is a significant bug that needs to be fixed as soon as possible. (I'm willing to fix it and create a pull request if you give the go ahead.)

https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html#torch.nn.LSTM

**Comment by sprout195** [comment]
i found the same question and failed to exploit the temperal characteristic of video when i use  ResnetCRNN, i fix the code as follow and i dont know whether is right:
RNN_out, (h_n, h_c) = self.LSTM(x_RNN.view(self.h_RNN,x_RNN.size(0),self.RNN_input_size), None)

**Comment by sohaib023** [comment]
@sprout195 

This will not work.
For example I have following array.

X = [[1, 2, 3, 4, 5], [6, 7, 8, 9, 10]]
X.view(5, 2) will yield
[[1, 2],
[3, 4],
[5, 6],
[7, 8],
[9,10]]

X.transpose(0,1) will yield
[[1, 6],
[2, 7],
[3, 8],
[4, 9],
[5, 10]]

Both yield the same shape but first one is erroneous. 
What you need is the latter.

**Comment by sohaib023** [comment]
For example, this code should work:

x_RNN = x_RNN.transpose(0, 1)
RNN_out, (h_n, c_n) = self.LSTM(x_RNN, None)
RNN_out = RNN_out.transpose(0, 1)

**Comment by sprout195** [comment]
it seem that the later code is right in theory but it dosen't work in my job, maybe some else solution?

**Comment by sprout195** [comment]
i try the later code（transpose） and it seem that the architecture lose its function.  maybe there is someelse solution

**Comment by sohaib023** [comment]
@sprout195 
Hey,
Could you please explain what exactly you mean by "lose its function"?
Is it giving an error or is it not learning?

**Comment by sprout195** [comment]
sorry.   lose it function means that it dosent work in classification or output error

**Comment by sprout195** [comment]
hello, maybe we make a mistake, look at the class ResCNNEncoder(nn.Module):         
cnn_embed_seq = torch.stack(cnn_embed_seq, dim=0).transpose_(0, 1)
here already have a transpose, and it means that there is no need to transpose in RNNdecoder

**Comment by sprout195** [comment]
hello, maybe we make a mistake, look at the class ResCNNEncoder(nn.Module):
cnn_embed_seq = torch.stack(cnn_embed_seq, dim=0).transpose_(0, 1)
here already have a transpose, and it means that there is no need to transpose in RNNdecoder

**Comment by sprout195** [comment]
@sohaib023 hey, i debug the code. the batch_size=36, seq_len=40 in my job. i found that the ouput tensor shape in Resnetencoder class: cnn_embed_seq shape is [36,40,512], 512 is the cnn embedding of single image input.  When the tensor input into the DecoderRNN. This is error. Moreover, x  = self.fc1(RNN_out[: , -1 , :]) exist error too. because the output of this tensor is choose the last batch but last time_step. i fix the code and hope your reply:


 def forward(self, x_RNN):  #x_RNN #shape: [36,40,512]
        self.LSTM.flatten_parameters()
        x_RNN = x_RNN.transpose_(0, 1) #x_RNN  shape:  [40,36,512]
        RNN_out, (h_n, h_c) = self.LSTM(x_RNN, None)
        """ h_n shape (n_layers, batch, hidden_size), h_c shape (n_layers, batch, hidden_size) """
        """ None represents zero initial hidden state. RNN_out has shape=(batch, time_step, output_size) """
        # FC layers
        x = self.fc1(RNN_out[-1, : , :])  # choose RNN_out at the last time step
        x = F.relu(x)                        #  fc1 layer output shape x: [36,256]

**Comment by sprout195** [comment]
we may make mistake. when the parameter batch_first of lstm is true.  it will has batch size as first dimension. (batch.timestep .input size)

**Comment by sohaib023** [comment]
@sprout195 

Hey, you're right. The original code was correct. I had overlooked the "batch_first" parameter.
Good find there :D I had given up and was looking to try some other approach like 3D CNNs instead.
However, I must say, it's not a very advisable feature to allow this on pytorch's end as altering input/output behavior will lead to confusion (as it did). Least they could do is mention a note in the input output descriptions about the "batch_first" flag.

Anyway. I'll close this issue as it has been resolved.
